File: clients/rn/src/engine/persistence/EventJournal.ts
/**
 * Event journal for room event sourcing.
 * TypeScript port of KMP's EventJournal.kt.
 */

import type { WorldEvent } from '../events/WorldEvent';

export interface EventJournal {
  append(roomId: string, event: WorldEvent): Promise<void>;
  replay(roomId: string): Promise<WorldEvent[]>;
  saveSnapshot(roomId: string, snapshotJson: string): Promise<void>;
  loadSnapshot(roomId: string): Promise<string | null>;

  /**
   * Compact events for a room, keeping only the most recent N events.
   * @param roomId  Room to compact.
   * @param keepLast  Number of events to retain (default 500).
   */
  compact(roomId: string, keepLast?: number): Promise<void>;

  /**
   * Return the number of persisted events for a room (diagnostics).
   */
  getEventCount(roomId: string): Promise<number>;
}

File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/persistence/EventJournal.kt
package org.wyrdsekai.app.engine.persistence

import org.wyrdsekai.app.engine.event.WorldEvent

/**
 * Event journal for room event sourcing.
 * Platform-specific implementations: SQLite on iOS, stubs on desktop/android
 * (those platforms run the full server subprocess instead).
 *
 * Snapshot methods use serialized JSON strings to avoid circular dependency
 * with RoomState (which depends on WorldEvent from this package's sibling).
 */
interface EventJournal {
    suspend fun append(roomId: String, event: WorldEvent)
    suspend fun replay(roomId: String): List<WorldEvent>
    suspend fun saveSnapshot(roomId: String, snapshotJson: String)
    suspend fun loadSnapshot(roomId: String): String?

    /**
     * Compacts the journal for a room, keeping only the most recent [keepLast] events.
     * Default implementation is a no-op for backward compatibility.
     */
    suspend fun compact(roomId: String, keepLast: Int = 500) { /* no-op default */ }
}
